yarn-project/merkle-tree/src/snapshots/append_only_snapshot.ts

```diff
@@ -0,0 +1,179 @@
+    const filledLeavesAtVersion = await this.#getLeafCountAtVersion(version);
+
+    if (typeof filledLeavesAtVersion === 'undefined') {
+      throw new Error(`Version ${version} does not exist for tree ${this.tree.getName()}`);
+    }
+
+    return new AppendOnlySnapshot(this.db, version, filledLeavesAtVersion, this.tree, this.hasher);
+  }
+
+  async snapshot(version: number): Promise<SiblingPathSource> {
+    const leafCountAtVersion = await this.#getLeafCountAtVersion(version);
+    if (typeof leafCountAtVersion !== 'undefined') {
+      throw new Error(`Version ${version} of tree ${this.tree.getName()} already exists`);
+    }
+
+    const batch = this.db.batch();
+    const root = this.tree.getRoot(false);
+    const depth = this.tree.getDepth();
+    const treeName = this.tree.getName();
+    const queue: [Buffer, number, bigint][] = [[root, 0, 0n]];
+
+    // walk the BF and update latest values
+    while (queue.length > 0) {
+      const [node, level, index] = queue.shift()!;
+
+      const previousValue = await this.db.get(nodePreviousValueKey(treeName, level, index)).catch(() => undefined);
+      if (!previousValue || !node.equals(previousValue)) {
+        // console.log(`Node at ${level}:${index} has changed`);
+        batch.put(nodeVersionKey(treeName, level, index), String(version));
+        batch.put(nodePreviousValueKey(treeName, level, index), node);
+      } else {
+        // if this node hasn't changed, that means, nothing below it has changed either
+        continue;
+      }
+
+      if (level + 1 > depth) {
+        // short circuit if we've reached the leaf level
+        // otherwise getNode might throw if we ask for the children of a leaf
+        continue;
+      }
+
+      const [lhs, rhs] = await Promise.all([
+        this.tree.getNode(level + 1, 2n * index),
+        this.tree.getNode(level + 1, 2n * index + 1n),
+      ]);
+
+      if (lhs) {
+        queue.push([lhs, level + 1, 2n * index]);
+      }
+
+      if (rhs) {
+        queue.push([rhs, level + 1, 2n * index + 1n]);
+      }
+    }
+
+    const leafCount = this.tree.getNumLeaves(false);
+    batch.put(snapshotMetaKey(treeName, version), leafCount);
+    await batch.write();
+
+    return new AppendOnlySnapshot(this.db, version, leafCount, this.tree, this.hasher);
+  }
+
+  async #getLeafCountAtVersion(version: number): Promise<bigint | undefined> {
+    const filledLeavesAtVersion = await this.db
+      .get(snapshotMetaKey(this.tree.getName(), version))
+      .then(x => BigInt(x.toString()))
+      .catch(() => undefined);
+    return filledLeavesAtVersion;
+  }
+}
+
+/**
+ * a
+ */
+class AppendOnlySnapshot implements SiblingPathSource {
+  constructor(
+    private db: LevelUp,
+    private version: number,
+    private leafCountAtVersion: bigint,
+    private tree: TreeBase & AppendOnlyTree,
+    private hasher: Hasher,
+  ) {}
+
+  public async getSiblingPath<N extends number>(index: bigint, _: boolean): Promise<SiblingPath<N>> {
```

Comment:
What is the `_: boolean` in here? What is its purpose?

Reply:
It comes from the `SiblingPathSource` interface and it's whether to include uncommitted data in the path. Oh I could even remove the parameter and TS should still be happy :thinking: